taf/updater/updater_pipeline.py

```diff
@@ -0,0 +1,1076 @@
+                            )
+
+                        taf_logger.info(
+                            f"Repository {repository.name}: found commits succeeding the last authenticated commit on branch {branch}: {','.join(additional_commits)}"
+                        )
+
+                        # these commits include all commits newer than last authenticated commit (if unauthenticated commits are allowed)
+                        # that does not necessarily mean that the local repository is not up to date with the remote
+                        # pull could've been run manually
+                        # check where the current local head is
+                        branch_current_head = repository.top_commit_of_branch(branch)
+                        if branch_current_head in additional_commits:
+                            additional_commits = additional_commits[
+                                additional_commits.index(branch_current_head) + 1 :
+                            ]
+                    self.state.additional_commits_per_target_repos_branches[
+                        repository.name
+                    ][branch] = additional_commits
+            return self.state.update_status
+        except UpdateFailedError as e:
+            self.state.error = e
+            taf_logger.error(e)
+            self.state.event = Event.PARTIAL
+            return UpdateStatus.PARTIAL
+        except Exception as e:
+            self.state.error = e
+            taf_logger.error(e)
+            self.state.event = Event.FAILED
+            return UpdateStatus.FAILED
+
+    @log_on_start(
+        INFO, "Merging commits into target repositories...", logger=taf_logger
+    )
+    def merge_commits(self):
+        """Determines which commits needs to be merged into the specified branch and
+        merge it.
+        """
+        try:
+            if self.only_validate:
+                return self.state.update_status
+
+            if self.state.update_status == UpdateStatus.FAILED:
+                # couldn't validate the first new commit
+                # there is nothing to merge
+                # remove cloned repositories if the initial commit was incorrect
+                for repository in self.state.cloned_target_repositories:
+                    taf_logger.debug("Removing cloned repository {}", repository.path)
+                    shutil.rmtree(repository.path, onerror=on_rm_error)
+            else:
+                for repository in self.state.target_repositories.values():
+                    # this will only include branches that were, at least partially, validated (up until a certain point)
+                    for (
+                        branch,
+                        validated_commits,
+                    ) in self.state.validated_commits_per_target_repos_branches[
+                        repository.name
+                    ].items():
```

Comment:
If I understand correctly validation is done BFS, but merging is DFS. Is it possible that during DFS merge a target repository merge fails (because of some conflict)? If so, will the next time we run updater, the target repository state affect target validation? Since one target repo would have multiple merged commits but other targets wouldn't.

Reply:
I did not modify the merge logic. If, for any reason, the state of one or more target repositories is not synced with the authentication repository's state, the validation should start from scratch: https://github.com/openlawlibrary/taf/blob/cd24900f3a0e05f4cb6dc45922c66e1cd679b4f5/taf/updater/updater_pipeline.py#L366